pulsar-io/hbase/src/main/java/org/apache/pulsar/io/hbase/sink/HbaseAbstractSink.java

```diff
@@ -0,0 +1,193 @@
+
+    protected HbaseUtils.TableDefinition tableDefinition;
+
+    // for flush
+    private List<Record<T>> incomingList;
+    private List<Record<T>> swapList;
+    private AtomicBoolean isFlushing;
+    private int timeoutMs;
+    private int batchSize;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        hbaseSinkConfig = HbaseSinkConfig.load(config);
+        if (hbaseSinkConfig.getZookeeperQuorum() == null
+                || hbaseSinkConfig.getZookeeperClientPort() == null
+                || hbaseSinkConfig.getTableName() == null) {
+            throw new IllegalArgumentException("Required property not set.");
+        }
+
+        getTable(hbaseSinkConfig);
+        tableDefinition = getTableDefinition(hbaseSinkConfig);
+
+        timeoutMs = hbaseSinkConfig.getTimeoutMs();
+        batchSize = hbaseSinkConfig.getBatchSize();
+
+        incomingList = Lists.newArrayList();
+        swapList = Lists.newArrayList();
+        isFlushing = new AtomicBoolean(false);
+
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), timeoutMs, timeoutMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        admin.close();
+        connection.close();
+    }
+
+    @Override
+    public void write(Record<T> record) throws Exception {
+        int number;
+        synchronized (incomingList) {
+            incomingList.add(record);
+            number = incomingList.size();
+        }
+
+        if (number == batchSize) {
+            flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    // bind value with a PreparedStetement
+    public abstract List<Put> bindValue(Record<T> message) throws Exception;
+
+    private void flush() {
+        // if not in flushing state, do flush, else return;
+        if (incomingList.size() > 0 && isFlushing.compareAndSet(false, true)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Starting flush, queue size: {}", incomingList.size());
+            }
+            if (!swapList.isEmpty()) {
+                throw new IllegalStateException("swapList should be empty since last flush. swapList.size: " + swapList.size());
+            }
+
+            synchronized (incomingList) {
+                List<Record<T>> tmpList;
+                swapList.clear();
+
+                tmpList = swapList;
+                swapList = incomingList;
+                incomingList = tmpList;
+            }
+            try {
+                // bind each record value
+                for (Record<T> record : swapList) {
+                    List<Put> puts = bindValue(record);
+                    if (CollectionUtils.isNotEmpty(puts)) {
+                        table.put(puts);
+                    }
+                    record.ack();
```

Comment:
it seems that we are acking before flushing. so this connector is an at-most-once connector? if it is an at-most-once connector, we should document it.

Reply:
> it seems that we are acking before flushing. so this connector is an at-most-once connector? if it is an at-most-once connector, we should document it.

use hbase put, so not have to at-most-once, I modify this